tests/test_tokenization_common.py

```diff
@@ -593,7 +593,10 @@ def test_pickle_added_tokens(self):
 
     def test_added_tokens_do_lower_case(self):
         # TODO(thom) activate fast tokenizer tests once Rust tokenizers accepts white spaces in added tokens.
-        tokenizers = [self.get_tokenizer(do_lower_case=True)] if self.test_slow_tokenizer else []
+        if not self.test_slow_tokenizer:
+            self.skipTest("This test is only for slow tokenizers")
+            return
```

Comment:
This would be consistent as line 1670.

Reply:
From the change you propose in the `test_encode_decode_with_spaces` test, I understand that rust tokenizers now accept spaces in added tokens. 

If this is the case, perhaps we should take the opportunity to modify this test to test the Rust Tokenizers as well (as suggested in the comment above). What do you think? :slightly_smiling_face:

Reply:
Sure, that makes sense. I'll work on it.

Reply:
After doing some tests, although the Rust tokenizer supports spaces in added tokens, two other problems, which seem unrelated to having spaces or not, raise:

The first one looks like that Rust tokenizer doesn't lowercase added tokens so the first assertion failed:
```
self = <tests.test_tokenization_albert.AlbertTokenizationTest testMethod=test_added_tokens_do_lower_case>

    def test_added_tokens_do_lower_case(self):
        # TODO(thom) activate fast tokenizer tests once Rust tokenizers accepts white spaces in added tokens.
        #if not self.test_slow_tokenizer:
        #    self.skipTest("This test is only for slow tokenizers")
        #    return
        tokenizers = self.get_tokenizers(fast=True, do_lower_case=True)
        for tokenizer in tokenizers:
            with self.subTest(f"{tokenizer.__class__.__name__}"):
                #if not hasattr(tokenizer, "do_lower_case") or not tokenizer.do_lower_case:
                #    continue
    
                special_token = tokenizer.all_special_tokens[0]
    
                text = special_token + " aaaaa bbbbbb low cccccccccdddddddd l " + special_token
                text2 = special_token + " AAAAA BBBBBB low CCCCCCCCCDDDDDDDD l " + special_token
    
                toks0 = tokenizer.tokenize(text)  # toks before adding new_toks
    
                new_toks = ["aaaaa bbbbbb", "cccccccccdddddddd", "AAAAA BBBBBB", "CCCCCCCCCDDDDDDDD"]
                added = tokenizer.add_tokens(new_toks)
>               self.assertEqual(added, 2, tokenizer.get_added_vocab())
E               AssertionError: 4 != 2 : {'CCCCCCCCCDDDDDDDD': 30003, 'cccccccccdddddddd': 30001, 'aaaaa bbbbbb': 30000, 'AAAAA BBBBBB': 30002}
```

The second one seems related to the special handling of mask token in `AlbertTokenizerFast`, resulting in the mask token not treated as a special token:

https://github.com/huggingface/transformers/blob/62832c962f85b5a554ebf8b930d13b76b9028a8d/src/transformers/models/albert/tokenization_albert_fast.py#L139

```
                # Check that none of the special tokens are lowercased
                sequence_with_special_tokens = "A " + " yEs ".join(tokenizer.all_special_tokens) + " B"
                tokenized_sequence = tokenizer.tokenize(sequence_with_special_tokens)
    
                for special_token in tokenizer.all_special_tokens:
>                   self.assertTrue(special_token in tokenized_sequence, f"{tokenizer.all_special_tokens} {tokenized_sequence}")
E                   AssertionError: False is not true : ['[CLS]', '[SEP]', '<unk>', '<pad>', '[MASK]'] ['▁a', '▁', '[CLS]', '▁yes', '▁', '[SEP]', '▁yes', '▁', '<unk>', '▁yes', '▁', '<pad>', '▁yes', '▁[', 'mask', ']', '▁b']
```

Reply:
Thanks a lot for the detail of the errors that prevent to extend this test to fast tokenizers.

I totally agree with your first analysis. Indeed, slow and fast tokenizers do not add in the same way the new tokens when `do_lower_case` is `True`. We should see on our side if it's something we want to harmonize or not (cc @sgugger and @LysandreJik ).

Regarding the second error, I think it's because we do `mask_token = AddedToken(mask_token, lstrip=True, rstrip=False)` and not `mask_token = AddedToken(mask_token, lstrip=True, rstrip=False, normalized=False)`. Did you have something else in mind?

Reply:
This is a very weird edge case but since texts will be lowercased, the special tokens added in fast tokenizers should be lowercased as well or they will never be identified in texts.

Reply:
@sgugger 
> This is a very weird edge case but since texts will be lowercased, the special tokens added in fast tokenizers should be lowercased as well or they will never be identified in texts.
I think that currently with the fast tokenizers it would be necessary to make wrap it into an `AddedToken` and set the `normalized` argument to `False` so that the special token is added to the tokenizer and well detected during the tokenization if it is present in the initial text. 

The difference with the slow tokenizer is that it will make the match before the normalization. For example:
```python
special_token =  '[SPECIAL TOKEN]'
tokenizer_fast = AutoTokenizer.from_pretrained("bert-base-uncased")
tokenizer_slow = AutoTokenizer.from_pretrained("bert-base-uncased", use_fast=False)

tokenizer_fast.add_tokens(AddedToken(special_token, normalized=False))
tokenizer_slow.add_tokens(special_token)

print(tokenizer_fast.convert_ids_to_tokens(tokenizer_fast.encode(special_token)))
print(tokenizer_slow.convert_ids_to_tokens(tokenizer_slow.encode(special_token)))
```
will output:
```
['[CLS]', '[SPECIAL TOKEN]', '[SEP]']
['[CLS]', '[special token]', '[SEP]']
```

Reply:
Oh understood! That seems like the right behavior for fast tokenizers! It is slightly different than for the slow one, but that's okay in this instance.

Reply:
Hi @SaulLu,

As PR #13930 was merged, the remaining problem would be the first one i.e. Rust tokenizers don't lowercase added tokens when `do_lower_case` is set to `True`, do you have any decision in mind to cope with this issue?

Reply:
Thanks again! For the first problem, what do you think about a test that distinguishes between slow and fast tokenizers, like what is done in [this test](https://github.com/huggingface/transformers/blob/v4.12.3/tests/test_tokenization_common.py#L953-L972) for example?

Reply:
Makes sense! There are two prerequisite PRs ( #14364 and #14365) awaiting merged to complete this test case.